src/v/cluster/topic_recovery_validator.cc

```diff
@@ -0,0 +1,307 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "topic_recovery_validator.h"
+
+#include "cloud_storage/anomalies_detector.h"
+#include "cluster/logger.h"
+
+#include <seastar/coroutine/as_future.hh>
+
+namespace cluster {
+
+// Each partition gets a separate retry_chain_node and a logger tied to it.
+// From this a common retry_chain_logger is created
+partition_validator::partition_validator(
+  cloud_storage::remote& remote,
+  cloud_storage_clients::bucket_name const& bucket,
+  ss::abort_source& as,
+  model::ntp ntp,
+  model::initial_revision_id rev_id,
+  recovery_checks checks)
+  : remote_{&remote}
+  , bucket_{&bucket}
+  , as_{&as}
+  , ntp_{std::move(ntp)}
+  , rev_id_{rev_id}
+  , op_rtc_{retry_chain_node{
+      as,
+      300s,
+      config::shard_local_cfg().cloud_storage_initial_backoff_ms.value()}}
+  , op_logger_{retry_chain_logger{
+      clusterlog,
+      op_rtc_,
+      ssx::sformat("recovery validation of {}/{}", ntp_, rev_id_)}}
+  , checks_{checks} {}
+
+ss::future<validation_result> partition_validator::run() {
+    op_logger_.debug(
+      "will perform checks: mode={}, max_segment_depth={}",
+      checks_.mode,
+      checks_.max_segment_depth);
+
+    switch (checks_.mode) {
+        using enum model::recovery_validation_mode;
+    case no_check:
+        // Handle this for switch-completeness, should be already
+        // handled by the caller
+        return ss::make_ready_future<validation_result>(
+          validation_result::passed);
+    case check_manifest_existence:
+        return do_validate_manifest_existence();
+    case check_manifest_and_segment_metadata:
+        return do_validate_manifest_metadata();
+    }
+
+    // Unhandled cases, compilation should prevent this. Ensure some
+    // level of forward compatibility
+    op_logger_.warn("{} not handled", checks_.mode);
+    return ss::make_ready_future<validation_result>(validation_result::passed);
```

Comment:
nit: could just use `__builtin_unreachable`? Regarding forward compatibility, wouldn't we fail to serialize the cluster property and never begin validation?